Create and test PolicyRepresentation

ZachParent opened the pull request:
## what changed
- split out `PolicyRepresentation` from `PolicyApproximator`
- reimplement core graph functionality from `PolicyGraph` in `GraphRepresentation`
  - use generic names like `state` and `transition`, as opposed to `node` and `edge`
  - add type checking
- add `test_policy_representation.py` with unit tests
- update `StateRepresentation` and `Action` in `discretizer.py`
- addresses #5 and #2 

## how it's tested
- unit tests are added to `test_policy_representation.py` (I recommend starting there when reviewing)

## notes
- some of the implementation details, I was unsure about. we should look closely at these later, especially as we integrate this into `PolicyApproximator`
- I left legacy methods like `nodes` and `edges` in `GraphRepresentation`, even though these are duplicates of the more general `states` and `transitions`. we could consider removing these already
- all type checks pass
- next steps:
  1. use this `PolicyRepresentation` in `PolicyApproximator`, when reimplementing `PolicyGraph` there
  2. update tests for `PolicyGraph` to point to `PolicyApproximator`

adriantormos reviewed (commented):
- Given that a new `StateRepresentation` class is being created to encapsulate the various possible state representations that will exist, doesn't it make more sense to leave it as an abstract class that serves as a common interface and put the logic in `PredicateBasedStateRepresentation`?
- In the future, actions may be more complex than an int serving as a categorical. I think it's important to be aware of it, but we'll actually cross that bridge when we need to.
- For the functions from `GraphRepresentation.NetworkXGraph` that return an iterator, is it possible to add the type of the elements that the iterator returns?
- In `GraphRepresentation.add_states_from()` and `GraphRepresentation.add_transitions_from()`, are the attributes that the user can pass to the function the same for all states or transitions that are added, or is the user supposed to pass a collection for each attribute? It seems a bit confusing, I don't see a 100% clear use for them yet.
- `GraphRepresentation.get_state_attributes()` and `GraphRepresentation.set_state_attributes()` assume the networkx representation is being used. They're probably too specific to be in `GraphRepresentation`.
- No comment on the minor edits to `PolicyGraph` given that it'll be re-implemented soon.

ZachParent commented:
> * Given that a new `StateRepresentation` class is being created to encapsulate the various possible state representations that will exist, doesn't it make more sense to leave it as an abstract class that serves as a common interface and put the logic in `PredicateBasedStateRepresentation`?

Yes! I totally agree. Updated.
> * In the future, actions may be more complex than an int serving as a categorical. I think it's important to be aware of it, but we'll actually cross that bridge when we need to.

Agreed. I did not have a good abstract idea of what an Action will need to look like, and env works fine with int actions, so I opted to not take an opinion for now. I've just added a `# TODO` so we keep track of this oversight.
> * For the functions from `GraphRepresentation.NetworkXGraph` that return an iterator, is it possible to add the type of the elements that the iterator returns?

Great point. These NetworkX methods return a specific type (specialized views, that implement Mapping and Set), and I've forwarded these types now.

ZachParent commented:
> * In `GraphRepresentation.add_states_from()` and `GraphRepresentation.add_transitions_from()`, are the attributes that the user can pass to the function the same for all states or transitions that are added, or is the user supposed to pass a collection for each attribute? It seems a bit confusing, I don't see a 100% clear use for them yet.

It's a good question, although I think this method is important, and it is essentially forwarding functionality from NetworkX. A good demonstration of how they are used is in the tests:
https://github.com/HPAI-BSC/pgeon/blob/ee29d54293f88a4110e6cb618672e40d74cc89e9/test/pgeon/test_policy_representation.py#L256-L258

it's also effectively used in PolicyGraph when initializing many nodes, like in `_update_with_trajectory()`: https://github.com/HPAI-BSC/pgeon/blob/4276aa246ac6363d1636f02c27ad38753d255d99/src/pgeon/policy_graph.py#L174-L182

and according to NetworkX documentation:
https://github.com/networkx/networkx/blob/0a58c919651de6b130391328776f84449c6fe784/networkx/classes/digraph.py#L489-L502

```
attr : keyword arguments, optional (default= no attributes)
            Update attributes for all nodes in nodes.
            Node attributes specified in nodes as a tuple take
            precedence over attributes specified via keyword arguments.
```

Let me know if you think a renaming / abstraction is appropriate, but this functionality seems important.

ZachParent commented:
> * `GraphRepresentation.get_state_attributes()` and `GraphRepresentation.set_state_attributes()` assume the networkx representation is being used. They're probably too specific to be in `GraphRepresentation`.

Hmmmm, I'm tempted to agree with you because they really forward NetworkX functionality directly, but I think this method call will be important in general, for example to update edges programmatically when normalizing:
https://github.com/HPAI-BSC/pgeon/blob/4276aa246ac6363d1636f02c27ad38753d255d99/src/pgeon/policy_graph.py#L200-L207

Maybe these method declarations don't belong in the highest level PolicyRepresentation, and more purpose-built methods like `normalize()` could be used. But I think that a way for programmatically  updating graph node/edge attributes is core to a GraphRepresentation. What do you think?

ZachParent commented:
Thank you for your feedback! I've responded to your comments 1 x 1, to keep conversations more manageable.

adriantormos reviewed (approved):
Some of the comments are still pending, but they can be left for later given the state of the project. We'll come back later to them and make a decision.
